scipio/src/sys/uring.rs

```diff
@@ -960,9 +951,21 @@ impl Reactor {
                     flush_rings!(lat_ring)?;
                 }
             }
-            if should_sleep {
-                self.link_rings_and_sleep(&mut main_ring)?;
+        }
+
+        if should_sleep {
+            if let Some(dur) = spin {
+                let start = Instant::now();
+                while start.elapsed() < dur {
+                    consume_rings!(into wakers; lat_ring, poll_ring, main_ring);
+
+                    if !wakers.is_empty() {
+                        return Ok(true);
```

Comment:
@glommer I am not sure if I need to rearm preempt timer before exiting here

Reply:
I don't know, but it got a bit hard to follow.
I think it will be more productive if you rework this a bit with your newfound understanding of what the preempt timer is given the recent explanations.

I think separating the sleepable and non-sleepable parts of wait is good (but then wait should be renamed to `poll`, or something else that doesn't give the idea that this is waiting on anything)

* For the poll part, we have to register a timer if there is a user timer specified.
* In the case where we sleep (or spun for a while, and stopped sleeping), the preempt timer should be registered when we return. Imagine if you spin for 1ms, and your preempt timer was 2ms before the spin: you already stole 1ms of this poor task queue! So only at the end.

The cleanest way is maybe for the `poll` function to never deal with the preempt timer, only the user timer.
The sleep function would always deal with the preempt timer, never the user timer.